Eric/movie ssr conversion

erc08 opened the pull request:
- Added SSR for fetching movie data
- Fixed page reload issues
- Restructured layout files (header, nav bar, footer)

erc08 committed:
fixed movie details page reloading issues using ssr

favorites button in movie dynamic route disabled due to conflicts between localstorage and SSR

erc08 committed:
Restructured files (#18)

erc08 committed:
Reformatting Search.tsx

reco commented on components/Layout/Layout.tsx:
```diff
@@ -0,0 +1,18 @@
+// Styles
+import styles from "./Layout.module.css";
+
+export default function Layout(props) {
+  return (
+    <div className={styles.container}>
+      <Header />
+      <Nav />
+      {props.children}
```
```suggestion
export default function Layout({children}) {
  return (
    <div className={styles.container}>
      <Header />
      <Nav />
      {children}
```

erc08 committed:
Layout optimizations

erc08 committed:
Moved Trending page to /movie, preparing for Favorites page

erc08 committed:
Fixed empty query error in Search, Restyled CSS

erc08 committed:
Update Layout.module.css

erc08 committed:
Restructuring search functionality using SSR

Preparing to move searchbar into navigation, allowing user to search from wherever

erc08 committed:
Major bugfix -- see description

- Favorites Button in Moviecard re-enabled.
- Fixed favorites button -- now works  on SSR pages, even when reloaded.
- Restructured SPA to have collapsed search bar built into NavBar, opens when clicked
- Removed unnecessary files (SHButton)
- Slight reformatting of homepage

erc08 committed:
Removed unnecessary parameter in useToggleFav

reco commented on components/FavoriteButton/FavoriteButton.tsx:
```diff
@@ -0,0 +1,52 @@
+// Libraries
+import React, { useState, useEffect, useMemo } from "react";
```
```suggestion
import React, { useState, useEffect } from "react";
```
if not needed remove

reco commented on components/FavoriteButton/FavoriteButton.module.css:
```diff
@@ -1,18 +1,17 @@
 .button {
-  color: black;
-  background-color: white;
+  color: var(--text-color);
```
great you are using css vars!

reco commented on components/FavoriteButton/FavoriteButton.tsx:
```diff
@@ -0,0 +1,52 @@
+  console.log(`- Movie [ ${id} ]: ${isFav} -`);
+
+  return (
+    hasMounted && (
```
any reason you want to render the button only on the client? the Dom should be in sync client and server... server can just use the default state. client then hydrates what ever state is in the local storage

reco commented on components/FavoriteButton/FavoriteButton.tsx:
```diff
@@ -0,0 +1,52 @@
+    hasMounted && (
+      <button
+        onClick={clickHandler}
+        className={cx([
+          styles.button,
+          styles[`variant__${variant}`],
+          { [styles.isFav]: isFav },
+        ])}
```
beautiful

reco commented on components/FavoriteButton/useToggleFav.tsx:
```diff
@@ -0,0 +1,47 @@
+import { useState } from "react";
```
this `useTobbleFav` does not match my code in `reco/#7-favorites-button`

erc08 replied:
I initially had your code, but I tweaked it slightly in terms of parameters and where some checks were.

reco commented on components/Layout/Footer.tsx:
```diff
@@ -0,0 +1,20 @@
+import styles from "./Layout.module.css";
+
+function Footer() {
+  console.log("Load Footer");
```
```suggestion
```

reco commented on components/Layout/Footer.tsx:
```diff
@@ -0,0 +1,20 @@
+  ); // End return
+}
+
+export default React.memo(Footer);
```
You can do this but this component is so small and simple almost not worth it. You might add more over had then performance gains. Good write up: 

https://felixgerschau.com/react-performance-react-memo/

reco commented on components/Layout/Header.tsx:
```diff
@@ -0,0 +1,36 @@
+import styles from "./Layout.module.css";
+
+function Header(/*  { pageName }: any  */) {
+  console.log("Load Header");
```
```suggestion
```

reco commented on components/Layout/Layout.tsx:
```diff
@@ -0,0 +1,17 @@
+// Components
+import Header from "./Header";
+import Nav from "./Nav";
+import Footer from "./Footer";
+
+// Styles
+import styles from "./Layout.module.css";
+
+export default function Layout({ children }) {
+  return (
+    <>
+      <Header />
+      <main className={styles.container}>{children}</main>
+      <Footer />
+    </>
+  );
+}
```
Nice!

reco commented on components/Layout/Nav.tsx:
```diff
@@ -0,0 +1,28 @@
+import styles from "./Layout.module.css";
+
+export default function Nav() {
+  console.log("Load Nav");
```
```suggestion
```
we never want to see a console.log in the code

reco commented on components/Layout/Nav.tsx:
```diff
@@ -0,0 +1,28 @@
+
+      <Search />
+
+      {/* <Link href="/search">
```
wanna bring this back?

erc08 replied:
nope!

reco commented on components/Movie/Movie.tsx:
```diff
@@ -1,128 +1,40 @@
 
 // Functions
-import { Favorites } from "components/Favorites";
+import Moviecard from "./Moviecard";
```
nice

reco commented on components/Trending/Trending.tsx:
```diff
@@ -1,58 +1,52 @@
-
-export default function Trending() {
+const api_key = "6c0d3427fcf17ddf80d04106c35a3a98";
+const url = `https://api.themoviedb.org/3/trending/movie/day?api_key=${api_key}`;
```
wanna put this into env variables

```
API_KEY: 
API_URL: 
```

https://nextjs.org/docs/basic-features/environment-variables

reco commented on components/Trending/Trending.tsx:
```diff
@@ -1,58 +1,52 @@
+
+    console.log("-- Loading Trending Movies --"); // Debugging purposes
+
+    try {
+      // Parse data and store into movies array
+      const res = await fetch(url);
+      const data = await res.json();
+      // console.log(data.results);
+      setMovies(data.results);
+    } catch (err) {
+      console.log(err);
+    }
+  };
```
wanna do this with `getServerSideProps `?

reco commented on components/Movie/Moviecard.tsx:
```diff
@@ -0,0 +1,96 @@
+  moviedata: string;
+}
+
+export default function Moviecard({ moviedata }: IMoviecardProps) {
```
Data should not come in here as a JSON string. Make sure you make a nice interface for movie.
Also Movie just renders `Movicard. What do you need `Movie` for?

reco commented on components/Movie/Moviecard.tsx:
```diff
@@ -0,0 +1,96 @@
+  ); // End Return
+}
+
+const months = [
+  undefined,
+  "January ",
+  "February ",
+  "March ",
+  "April ",
+  "May ",
+  "June ",
+  "July ",
+  "August ",
+  "September ",
+  "October ",
+  "November ",
+  "December ",
+];
+
+function parseReleaseDate(rawDate: string) {
+  let dateArr = rawDate.split("-");
+  let month = months[parseInt(dateArr[1])];
+  let date = month + parseInt(dateArr[2]).toString() + ", " + dateArr[0];
+  return date;
+}
```
you could move this to `/utils` and then import it

reco commented on styles/globals.css:
```diff
@@ -1,16 +1,37 @@
+:root {
+  --header-color: #202020;
+  --bg-color: #181818;
+  --border-color: #8899a6;
+  --card-color: #282828;
+  --text-color: #fafafa;
+  --alt-text-color: #aaaaaa;
+  --highlight-color: #2571ff;
+  --alt-highlight-color: #888888;
+  --fav-color: #ff5f5f;
+  --accent-color: #ffffff;
+}
+
```
cool

reco commented on components/Search/Search.tsx:
```diff
@@ -1,81 +1,38 @@
 // Functions
-import { Results } from 'components/Results';
-import { Trending } from 'components/Trending';
+import { Results } from "components/Results";
+import { Trending } from "components/Trending";
```
not used

reco commented on components/Search/Search.tsx:
```diff
@@ -1,81 +1,38 @@
-}
\ No newline at end of file
+import sbstyles from "./Searchbar.module.css";
+import styles from "./Search.module.css";
```
not used

reco commented on components/Search/Searchbar.tsx:
```diff
@@ -0,0 +1,81 @@
+import { useRouter } from "next/router";
+
+// Functions
+import { Results } from "components/Results";
```
not used

reco commented on components/Search/Searchbar.tsx:
```diff
@@ -0,0 +1,81 @@
+function Searchbar() {
+  const [query, setQuery] = useState("");
+
+  // const [movies, setMovies] = useState([]);
```
console.log

reco commented on components/Search/Searchbar.tsx:
```diff
@@ -0,0 +1,81 @@
+    } catch (err) {
+      console.error(err);
+    }
+  }; // End searchMovies */
```
?

reco commented on components/Search/index.tsx:
```diff
@@ -1 +1 @@
-export { default as Search } from './Search';
```
nice linter working?

reco commented on pages/search/[query].tsx:
```diff
@@ -0,0 +1,37 @@
+import { Results } from "components/Results";
+
+// API data
+const api_key = "6c0d3427fcf17ddf80d04106c35a3a98";
+const api_url = `https://api.themoviedb.org/3/search/movie?api_key=${api_key}`;
```
env variables

reco commented on pages/search/[query].tsx:
```diff
@@ -0,0 +1,37 @@
+
+export async function getServerSideProps(context: GetServerSidePropsContext) {
+  const { query } = context.params;
+  console.log(query);
```
console.log

reco commented on pages/search/[query].tsx:
```diff
@@ -0,0 +1,37 @@
+
+  const url = api_url.concat(`&query=${query}`);
+
+  console.log("-- Submitting Search Query --"); // Debugging purposes
```
console.log

reco commented:
hey @ericc0429 [public/searchIcon512px.png](https://github.com/ericc0429/moviebrowser/pull/19#diff-96dd2a3db0ab764446c621a45841519f4b63ee5a2925b996a589d747404f9757) an sag

erc08 committed:
Cleaned up files, added SSR for Trending

erc08 committed:
Hide API key with environment variables

erc08 committed:
Moved rest of api data into environment variables

erc08 committed:
Moved parseDate to /utils

erc08 committed:
cleaned code

erc08 committed:
Update useToggleFav.tsx

erc08 committed:
Generalizing MovieCard component

erc08 committed:
Added env variables

reco committed:
add a tag

erc08 committed:
Converted to MovieCards.  Results is now depreciated, to be removed in future update.

erc08 committed:
Fixed favorites button in link component

erc08 committed:
Restructured files, cleaned up

reco commented on components/MovieCard/IMovieData.tsx:
```diff
@@ -0,0 +1,57 @@
+interface IProdCompany {
```
- For interface declaration you do not need JSX. So change the extension from `tsx` to `ts`
- These are types not data. Call the file `MovieCard.types.ts`

reco commented on components/MovieCard/index.tsx:
```diff
@@ -0,0 +1,3 @@
+export { default as MovieCard } from "./MovieCard";
+export * from "./IMovieData";
+export * from "./MovieCardList";
```
don't you want to move MovieList to a new dir in components?

reco commented on pages/movie/[id].tsx:
```diff
@@ -2,14 +2,16 @@
 
 export default function Main({ data }) {
-  return <Movie moviedata={JSON.stringify(data)} />; // End return
+  console.log(data);
```
damn... 

```suggestion
```

reco commented on components/MovieCard/MovieCard.tsx:
```diff
@@ -0,0 +1,113 @@
+}
+
+export default function MovieCard({ movie, variant }: IMovieProps) {
+  console.log("Moviecard");
```
```suggestion
```
damn

reco commented on pages/search/[query].tsx:
```diff
@@ -2,13 +2,15 @@
 const api_url = process.env.API_URL_SEARCH;
 
 export default function Main({ data }) {
+  console.log(typeof data);
```
```suggestion
```
damn

erc08 committed:
Changed IMovieData to MovieCard.types.ts

reco commented on components/MovieCard/IMovieData.tsx:
```diff
@@ -17,6 +17,10 @@ interface IMovieGenre {
   name: string;
 }
 
+/* Movie Data Interface ==================
```
- For interface declaration you do not need JSX. So change the extension from tsx to ts
- These are types not data. Call the file `MovieCard.types.ts`

erc08 committed:
Update searchicon
